core/src/main/java/io/aiven/kafka/tieredstorage/fetch/index/MemorySegmentIndexesCache.java

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright 2023 Aiven Oy
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.aiven.kafka.tieredstorage.fetch.index;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ForkJoinPool;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.function.Supplier;
+
+import org.apache.kafka.common.config.ConfigDef;
+import org.apache.kafka.server.log.remote.storage.RemoteStorageManager.IndexType;
+
+import io.aiven.kafka.tieredstorage.config.CacheConfig;
+import io.aiven.kafka.tieredstorage.metrics.CaffeineStatsCounter;
+import io.aiven.kafka.tieredstorage.storage.ObjectKey;
+import io.aiven.kafka.tieredstorage.storage.StorageBackendException;
+
+import com.github.benmanes.caffeine.cache.AsyncCache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.github.benmanes.caffeine.cache.RemovalListener;
+import com.github.benmanes.caffeine.cache.Scheduler;
+import com.github.benmanes.caffeine.cache.Weigher;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class MemorySegmentIndexesCache implements SegmentIndexesCache {
+    private static final Logger log = LoggerFactory.getLogger(MemorySegmentIndexesCache.class);
+
+    private static final long GET_TIMEOUT_SEC = 10;
+    private static final long DEFAULT_MAX_SIZE_BYTES = 100 * 1024 * 1024;
+    private static final String METRIC_GROUP = "segment-indexes-cache";
+
+    private final Executor executor = new ForkJoinPool();
+    private final CaffeineStatsCounter statsCounter = new CaffeineStatsCounter(METRIC_GROUP);
+
+    protected AsyncCache<SegmentIndexKey, byte[]> cache;
+
+    public RemovalListener<SegmentIndexKey, byte[]> removalListener() {
+        return (key, content, cause) -> log.debug("Deleted cached value for key {} from cache."
+            + " The reason of the deletion is {}", key, cause);
+    }
+
+    public Weigher<SegmentIndexKey, byte[]> weigher() {
+        return (key, value) -> value.length;
+    }
```

Comment:
Do these two need to be higher-order functions?
```suggestion
    private static void removalListener(final SegmentIndexKey key, final byte[] value, RemovalCause cause) {
        log.debug("Deleted cached value for key {} from cache. The reason of the deletion is {}", key, cause);
    }

    private static int weigher(final SegmentIndexKey key, byte[] value) {
        return value.length;
    }
```